src/python/pants/goal/run_tracker.py

```diff
@@ -390,3 +413,40 @@ def shutdown_worker_pool(self):
     N.B. This exists only for internal use and to afford for fork()-safe operation in pantsd.
     """
     SubprocPool.shutdown(self._aborted)
+
+  def report_target_info(self, scope, target, keys, val):
+    """Add target information to run_info under target_data.
+
+    Will Recursively construct a nested dict with the keys provided.
```

Comment:
The idea behind this API is that tasks might need additional structure, but will be making multiple calls to this method, rather than batching things up like the test code is doing.

So rather than calling:
```
report_target_info(scope, target, ['this'], {'is': {'a': {'deeply': {'namespaced': 'value'}}}})
```
...they'd call:
```
report_target_info(scope, target, ['this', 'is', 'a', 'deeply', 'namespaced'], 'value')
```